File: backend/models/timeline.py
"""
Pydantic models for the Cosmic Timeline (काल प्रवाह) system.
Defines schemas for timeline events, Vimshottari Dasha periods,
planetary transits, festivals, eclipses, requests, and unified response payloads.
"""

from pydantic import BaseModel
from typing import List, Optional, Dict, Any


class TimelineEvent(BaseModel):
    """Canonical single timeline node/event across any layer."""
    eventId: str
    layer: str  # birth, dasha, transit, festival, eclipse, muhurat
    title: str
    sanskritTitle: str
    category: str
    date: str  # YYYY-MM-DD
    dateFormatted: str  # "August 15, 1998"
    timeStr: Optional[str] = None
    icon: str
    importance: str  # Supreme, High, Medium, Subtle
    badgeColor: str
    description: str
    panchangContext: Optional[str] = None
    chartConnection: Optional[str] = None
    isBookmarked: bool = False
    metadata: Dict[str, Any] = {}


class DashaPeriod(BaseModel):
    """Vimshottari Dasha or Antardasha period definition."""
    lord: str
    sanskritLord: str
    lordSymbol: str
    startDate: str  # YYYY-MM-DD
    endDate: str  # YYYY-MM-DD
    durationYears: float
    isCurrent: bool = False
    antardashas: Optional[List["DashaPeriod"]] = None


class TransitEvent(BaseModel):
    """Planetary transit or sign ingress event."""
    planet: str
    sanskritPlanet: str
    fromSign: str
    toSign: str
    ingressDate: str
    isRetrograde: bool = False
    description: str


class FestivalEvent(BaseModel):
    """Vedic festival item."""
    id: str
    name: str
    sanskritName: str
    date: str
    paksha: str
    tithi: str
    significance: str
    deity: str


class EclipseEvent(BaseModel):
    """Solar or Lunar eclipse item."""
    id: str
    type: str  # Solar, Lunar
    typeSanskrit: str
    date: str
    timeWindow: str
    sign: str
    nakshatra: str
    traditionalNotes: str


class TimelineRequest(BaseModel):
    """Request body for generating a unified Cosmic Timeline."""
    fullName: str = "Observer"
    dateOfBirth: str  # YYYY-MM-DD
    timeOfBirth: str  # HH:MM
    city: str = "Ujjain"
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    timezone: Optional[str] = None
    startDate: Optional[str] = None  # YYYY-MM-DD
    endDate: Optional[str] = None  # YYYY-MM-DD
    activeLayers: List[str] = ["birth", "dasha", "transit", "festival", "eclipse", "muhurat"]


class TimelineResponse(BaseModel):
    """Canonical unified Cosmic Timeline response payload."""
    reportId: str
    fullName: str
    birthDate: str
    birthLocation: str
    currentDate: str
    activeLayers: List[str]
    totalEventsCount: int
    currentDasha: Dict[str, Any]
    events: List[TimelineEvent]
    transparency: Dict[str, Any]

File: src/types/timeline.ts
/**
 * Kalachakra — Cosmic Timeline Types
 * ===================================
 * Type definitions for the Cosmic Timeline (काल प्रवाह) system.
 */

export interface TimelineEvent {
  eventId: string;
  layer: 'birth' | 'dasha' | 'transit' | 'festival' | 'eclipse' | 'muhurat';
  title: string;
  sanskritTitle: string;
  category: string;
  date: string; // YYYY-MM-DD
  dateFormatted: string;
  timeStr?: string;
  icon: string;
  importance: 'Supreme' | 'High' | 'Medium' | 'Subtle';
  badgeColor: string;
  description: string;
  panchangContext?: string;
  chartConnection?: string;
  isBookmarked?: boolean;
  metadata?: Record<string, any>;
}

export interface FestivalEvent {
  id: string;
  name: string;
  sanskritName: string;
  date: string;
  paksha: string;
  tithi: string;
  significance: string;
  deity: string;
}

export interface EclipseEvent {
  id: string;
  type: string;
  typeSanskrit: string;
  date: string;
  timeWindow: string;
  sign: string;
  nakshatra: string;
  traditionalNotes: string;
}

export interface TimelineRequest {
  fullName: string;
  dateOfBirth: string;
  timeOfBirth: string;
  city: string;
  latitude?: number;
  longitude?: number;
  timezone?: string;
  startDate?: string;
  endDate?: string;
  activeLayers?: string[];
}

export interface TimelineResponse {
  reportId: string;
  fullName: string;
  birthDate: string;
  birthLocation: string;
  currentDate: string;
  activeLayers: string[];
  totalEventsCount: number;
  currentDasha: {
    mahadasha: string;
    sanskritMahadasha: string;
    mahadashaRange: string;
    antardasha: string;
    sanskritAntardasha: string;
    antardashaRange: string;
  };
  events: TimelineEvent[];
  transparency: {
    ayanamsha: string;
    houseSystem: string;
    dashaSystem: string;
    generatedTimestamp: string;
    disclaimer: string;
  };
}
